core/src/main/java/software/amazon/awssdk/core/runtime/http/response/JsonResponseHandler.java

```diff
@@ -111,8 +111,15 @@ public T handle(HttpResponse response, ExecutionAttributes executionAttributes)
                 IoUtils.drainInputStream(response.getContent());
             }
 
+            Map<String, String> metadata = unmarshallerContext.getMetadata();
+            metadata.put(SdkResponseMetadata.AWS_REQUEST_ID,
+                         response.getHeaders().get(X_AMZN_REQUEST_ID_HEADER));
+
+            T resultWithMetadata = (T) result.toBuilder().sdkResponseMetadata(new SdkResponseMetadata(metadata)).build();
```

Comment:
Do we want to be in the business of extracting this information in the JSON handler? I'm a bit skeptical of a custom metadata object like this, since it couples us pretty strongly AWS.

Maybe we could add the metadata object in the AwsResponse level instead?

```Java
class AwsResponse {
   AwsResponseMetadata awsResponseMetadata() {
      return new AwsResponseMetadata(this);
   }
}

class AwsResponseMetadata {
    AwsResponseMetadata(AwsResponse response) {
        // extract metadata from response
    }
}
```